### PR TITLE
Fix fused add RMSNorm residual-sum variance

### LINKED ISSUE #52104: [Bug]: vllm_c fused_add_rms_norm rounds the residual sum before variance
### Summary

The CUDA `vllm_c` implementation of `fused_add_rms_norm` and the native vLLM IR implementation use different rounding orders for BF16/FP16 inputs.

At upstream commit [`903d2efe`](https://github.com/vllm-project/vllm/commit/903d2efe7eb6d045d03c0a8192b1ac7bec50d168):

- The [native IR implementation](https://github.com/vllm-project/vllm/blob/903d2efe7eb6d045d03c0a8192b1ac7bec50d168/vllm/ir/ops/layernorm.py#L37-L55) computes `x.float() + residual.float()` and uses that unrounded FP32 sum to calculate the variance.
- The [vectorized `vllm_c` CUDA kernel](https://github.com/vllm-project/vllm/blob/903d2efe7eb6d045d03c0a8192b1ac7bec50d168/csrc/libtorch_stable/layernorm_kernels.cu#L133-L169) stores the sum in an `_f16Vec<scalar_t, width>` before `sum_squares()`. This rounds the residual sum to BF16/FP16 before calculating the variance.
- The [generic CUDA kernel](https://github.com/vllm-project/vllm/blob/903d2efe7eb6d045d03c0a8192b1ac7bec50d168/csrc/libtorch_stable/layernorm_kernels.cu#L191-L215) has the same behavior because `z` is `scalar_t`.

This matters because CUDA selects `["native"]` with Inductor, but selects `["vllm_c", "native"]` when compilation is disabled ([dispatch code](https://github.com/vllm-project/vllm/blob/903d2efe7eb6d045d03c0a8192b1ac7bec50d168/vllm/platforms/cuda.py#L686-L700)). Therefore, changing only `CompilationMode` can change numerical results and, after accumulation through many layers, greedy output tokens.

I observed this with GLM-4.7-Flash in BF16: the `CompilationMode.NONE` path diverged from the compiled/native path. Making the CUDA kernel use the FP32 residual sum for variance removed the token divergence in that test.

### Minimal reproducer

Run this against unmodified vLLM `main` on a CUDA GPU:

```python
import torch
from vllm import _custom_ops as ops

torch.manual_seed(0)
device = "cuda"
dtype = torch.bfloat16
tokens = 4
hidden_size = 2048
epsilon = 1e-6

x = torch.randn(tokens, hidden_size, device=device, dtype=dtype)
residual = torch.randn_like(x)
weight = torch.randn(hidden_size, device=device, dtype=dtype)

# Native IR semantics: keep input + residual in FP32 through variance.
summed_fp32 = x.float() + residual.float()
variance_fp32 = summed_fp32.square().mean(dim=-1, keepdim=True)
native = summed_fp32 * torch.rsqrt(variance_fp32 + epsilon)
native = (native.to(weight.dtype) * weight).to(dtype)
native_residual = summed_fp32.to(dtype)

# vllm_c custom op mutates both tensors in place.
actual = x.clone()
actual_residual = residual.clone()
ops.fused_add_rms_norm(actual, actual_residual, weight, epsilon)
torch.cuda.synchronize()

diff = (actual.float() - native.float()).abs()
print("output mismatches:", torch.count_nonzero(actual != native).item())
print("max output diff:", diff.max().item())
print("mean output diff:", diff.mean().item())
print(
    "residual mismatches:",
    torch.count_nonzero(actual_residual != native_residual).item(),
)
```

The exact mismatch count can vary with the GPU/build, but the two implementations are structurally different: `vllm_c` calculates variance from the rounded low-precision sum, while native IR calculates it from the FP32 sum.

For scale, simulating the current `vllm_c` arithmetic on an A100 with PyTorch 2.11.0/CUDA 13.0 and the seed/shapes above produced:

```text
tokens=1:  368 / 2048 output elements differ; max abs diff 0.03125
tokens=4:  1465 / 8192 output elements differ; max abs diff 0.0625
tokens=16: 5791 / 32768 output elements differ; max abs diff 0.0625
```

### Expected behavior

The `vllm_c` and native implementations should use the same defined arithmetic/rounding semantics so that selecting `CompilationMode.NONE` does not introduce a different residual-normalization calculation.

### Proposed fix

For both CUDA kernel specializations:

1. Compute each residual sum as FP32:
   `float sum = float(input) + float(residual)`.
2. Accumulate variance from that FP32 `sum`.
3. Publish `residual = scalar_t(sum)`.
4. Normalize using the FP32 `sum`, while preserving the intended output/weight multiplication rounding policy.

This requires recomputing the sum during the second pass or temporarily retaining it; reading the already-rounded `residual` in the second pass preserves the current mismatch.

### Related issues

This appears distinct from:

- #49616 / #49639, which concern rounding the normalized value before multiplying by the RMSNorm weight.
- #42379, which concerns the dtype used for weight multiplication.
- #48271, which concerns batch-size-dependent reduction behavior.
- #43390, which concerns integer overflow.

This report is specifically about rounding `input + residual` before variance.

### Environment used during investigation

- GPU: NVIDIA A100-SXM4-80GB
- PyTorch: 2.11.0+cu130
- CUDA: 13.0
- Driver: 580.105.08
- Dtype: BF16
- Model-level observation: GLM-4.7-Flash
- Relevant configuration: `CompilationMode.NONE`

### Before submitting a new issue

- [x] I searched existing issues and PRs for this specific residual-sum rounding mismatch.

This issue was prepared with AI assistance; the source comparison and reported measurements were manually validated on the hardware above.

### PR DESCRIPTION
**Pull request by danielpolimac**
## Purpose

Fixes #52104.

The CUDA `vllm_c` fused-add RMSNorm kernel was computing variance from the low-precision rounded `input + residual` value. The native IR implementation keeps `input.float() + residual.float()` in FP32 through variance calculation, then rounds only the published residual output.

This PR aligns the vectorized and generic `fused_add_rms_norm` CUDA kernel paths with that native arithmetic contract:

- First pass computes variance from the FP32 residual sum without mutating `residual`.
- Second pass recomputes the FP32 sum, writes the rounded residual output, and normalizes from the FP32 sum.
- Adds a focused `vllm_c` regression test that checks the output is closer to the FP32-sum contract than to the old rounded-sum contract.

Duplicate-work checks run before opening this PR:

- `gh issue view 52104 --repo vllm-project/vllm --comments`: issue open; one comment suggesting regression coverage.
- `gh pr list --repo vllm-project/vllm --state open --search "52104 in:body"`: no open PRs.
- `gh pr list --repo vllm-project/vllm --state open --search "fused_add_rms_norm residual sum variance BF16 FP32 rounding"`: no open PRs.

AI assistance was used to investigate and prepare this change. The submitting human should review and be able to defend every changed line before merge.

## Test Plan

- Syntax check touched Python test file.
- Focused pytest selection for the new regression test.
- Ruff format/check on the touched Python test file.
- Clang-format on the touched CUDA file.
- Whitespace diff check.

No model evals were run; this is a kernel arithmetic consistency fix, not a model-output-quality change. Full CUDA validation should be run in CI because the local host has no GPGPU.

## Test Result

- `./.venv/bin/python -m py_compile tests/kernels/ir/test_layernorm.py`: passed.
- `./.venv/bin/python -m pytest tests/kernels/ir/test_layernorm.py -k fp32_sum_for_variance -v`: selected 2 tests; both skipped locally because this macOS host has no GPGPU.
- `./.venv/bin/pre-commit run ruff-format --files tests/kernels/ir/test_layernorm.py`: passed.
- `./.venv/bin/pre-commit run ruff-check --files tests/kernels/ir/test_layernorm.py`: passed.
- `./.venv/bin/pre-commit run clang-format --files csrc/libtorch_stable/layernorm_kernels.cu`: passed.
- `git diff --check`: passed.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [x] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

**Comment by danielpolimac** [comment]
Closing this to avoid duplicate work under AGENTS.md now that #52243 exists for the same issue.

#52243 covers the same first-pass/second-pass variance fix and also points out a weighted-output rounding detail where this PR does not exactly match the native IR arithmetic. It has a stricter CUDA regression test for that contract, so it is the better path for #52104.

Thanks for the review bandwidth; I will leave this closed unless maintainers explicitly prefer reviving this branch.
